### ISSUE TITLE
Option to disable external link previews (globally as a reader, and per-toor as an author)

### ISSUE THREAD

**Issue by naught101** [mastodon/mastodon#20188]
### Pitch

External links often appear in the timeline as large cards that can contain some image or video content, e.g.:

![image](https://user-images.githubusercontent.com/167164/200704974-456afb12-0b4a-4031-bc52-ab9c7cec214b.png)

These cards are big and disruptive, and are sometimes annoying to have on your general feed.

It would be good if there was:

- An option to disable previews for all viewed toots in the interface (as a reader).
- An option to disable previews for each toot posted (as an author).

### Motivation

Some users are using Content Warnings as a way to hide these previews, but this devalues the CW as a way to warn people about potentially triggering content (as you don't necessarily know what the reason for the CW is).

Having the option to hide these per-user and per-toot would make CWs more useful, and would allow the feed to be kept cleaner, and make the interface slightly faster in some cases.

**Comment by afontenot** [comment]
Ideally, this would include a way to treat opengraph previews seperately from oembed embeds. The latter are almost always pretty reasonable and static, and most users probably won't want to disable them. Embeds are more disruptive and require a click because the content is from an external site. (The YouTube example shown in the screenshot above is an oembed.)

**Comment by naught101** [comment]
Is there a difference from the perspective of a user posting a link? Or is the type of preview that gets shown only dependent on the external website?

**Comment by afontenot** [comment]
No, there's no difference from the perspective of the user. It's up to the site. Basically, sites with news articles often go with the opengraph preview because it's just a nice little card with an image that shows you something about the article's contents. Sites with embeddable media like gif sites, video sites, etc, will use oembed, because the idea is that you should be able to view the media directly on the page.

**Comment by rbairwell** [comment]
As mentioned on the Discord, I think an option to "Disable link previews"  also brings additional security/privacy related enhancements. Say an infosec related instance is talking about a particular "bad actor" site - the last thing they want is for the site to be notified with a user agent like "http.rb/5.0.4 (Mastodon/1.0.7+3.5.5; +https://example.com/) Bot" with the reverse IP address.

**Comment by naught101** [comment]
Thunderbird by default disables external media in messages, and shows a banner that tells you this, and then provides options to allow the link preview now, and also to permanently add any domains involved to a whitelist:

![image](https://user-images.githubusercontent.com/167164/204163065-63a1852b-9741-4c17-91e6-713e66118bd3.png)

This might be a bit of overkill per-user for mastodon, but a whitelist for instance admins might be useful. This should be separate from the user-level option to disable previews though.

**Comment by MarcGuay** [comment]
Upvoting for on a different reason:  My website has cheap shared hosting and when I post a link to it on Mastodon the server gets crushed by all of the traffic crawling it to fetch the preview.  Being able to disable would help that.

**Comment by pamela-drouin** [comment]
Adding my 2 cents to this -- I agree. I wish I could turn off the external link previews. I find the experience of the preview unfurling in my feed very distracting and disorienting. I tend not to use the web version of Mastodon for this very reason and resort to my mobile app of choice instead.

**Comment by nielsle** [comment]
I would personally prefer a single option to turn of all link previews in all posts.

I don't need special settings for each author. Keep it simple :-)

**Comment by wioxjk** [comment]
I would personally love a way to disable previews. Since 70% of all my sidekiq jobs is "LinkCrawlWorker". I think this would reduce the load on the instance by alot.

**Comment by and-rea** [comment]
Server preview is a mixed issue:
on one hand. it serves to gather "good" info from the linked page, but
on the other hand, hammers the website with a flood of requests causing the owner one or a mix of the following:
- additional billing for traffic;
- momentary ddos for exceeded capacity;
- (even worse, for the family/little web owners) ddos for all the rest of the month because reaching monthly allowed bandwidth.

The option of enabling/disabling it, is not a solution because some instances/users will do, some others don't, and the original problem will persists even after the option exists.

I'm for a more simple approach, done in two steps:
1) all instances no more fetching link previews, and
2) **only the instance of the posting user will fetch the preview**, and attach it as an image to the user's post, that will be handled as every attached image of a post.

**Comment by jocelynetienne** [comment]
Another reason to allow per-user (or per-toot) link "preview card" or at least the preview image: bots posting links but willing to have as low an impact on server load as possible.

Example: I manage bots posting links to new scientific research papers or preprints. Some servers hosting these (eg arXiv, PubMED) use a huge PNG of their logo as oembed image. As far as I understand, that image is totally uninformative (always the same one) and gets duplicated with every toot times every server on which there are followers of the bot.

**Comment by emberfiend** [comment]
Sticking `mastodon.online##.status-card` in your uBlock Origin filter list handles the empty mini-link previews which give you the same information as the first 40 characters of the link plus the title usually already included in the toot. 

However, (this is a guess) once the scraper finds an image at the link they seem to turn into `media-gallery`s, which are also used for poster-uploaded images.

This sucks because a giant gallery image using 50% viewport height for a random logo from the linked site is pointless 90% of the time. But if you hide `media-gallery` you just lose all user-posted images. I would love to be able to hide `media-gallery` for a scraped image from a link but not for poster-uploaded images.

Finally there's also `video-player` for videos.

All of which is to say, yeah, Mastodon needs better control over link preview visibility.

**Comment by Diwannist** [comment]
We think it might be good, as readers, to have an option to hide link previews in _Preferences_.

**Comment by quicoto** [comment]
Would it be possible to have a server setting to disable Preview Cards entirely?

Which would also save space on the database and disk.

**Comment by ecstasyvies** [comment]
I’m also having a problem with Mastodon’s **rich preview** when I share links from my [calendar](https://multiplasidentidades.nekoweb.org/calendario/) that have specific event query strings (like `?evento=dia-da-lembranca-intersexo`).

Right now, Mastodon makes **two links**:  
- the **direct link** to the event  
- a **rich preview** based on the page’s metadata (OG tags)  

This ends up duplicating the link and causing weird behavior.

Ideally, the rich preview should just use the **direct event link**, or there should be a way to **turn off rich previews** for URLs like this.

Fixing this would make sharing links from **dynamic services**, like calendars or event systems, more predictable and useful, and keep the timeline **clean and clear**.

**Comment by MadokVaur** [comment]
> 
> Right now, Mastodon makes **two links**:
> 
>     * the **direct link** to the event
> 
>     * a **rich preview** based on the page’s metadata (OG tags)
> 
> 
> This ends up duplicating the link and causing weird behavior.

There is more 'weird behavior' than this

Often the metadata of the linked page generates a preview image that has *nothing* to do with the subject of the Mastodon post, and the text pulled over with the preview image is also a non sequitur to the text body of the Mastodon post itself

Then, the link behind the link preview image often overrides the direct text link provided by the poster, causing those who spontaneously click on images, expecting to be taken to the same destination the writer of the text link intended, are not linked to the subject of the Mastodon post

Particularly, this often overrides "free link" URLs

Historically I've forgotten exactly why the preview cards seemed to be a good idea to someone (who shall remain unnamed) in the distant past, but the effect has been consistently pretty intrusive, if not simply malformed, in the way Preview Cards actually work in the real world

**Comment by sorz** [comment]
As a mastodon user (who post), sometimes I'm intended to share a link, I'd like to have a preview card as normal. However, in other times, I may prefer to not have a preview card, when - 

1. I attached links as footnotes/references/FYI. I expect most people don't want click that links but still want provide them just in case. A prominent picture shown below my post feels distraction.
2. Some websites just use their logo/banner as preview card, with little information about that particular page. 

I like the Telegram-way. When I paste link to the edit box, there is a "loading preview card" with delete button shown along with the edit box. I can either click "delete" before preview card loaded (case 1), or after (case 2) to disable preview card for that post.

**Comment by trwnh** [comment]
cross-referencing https://w3id.org/fep/8967

**Comment by Kpt-Kvern** [comment]
Mastodon offers an Appearance option to Always hide media, in the Sensitive content sub-category. This option doesn't hide media showed in PreviewCards.

From an accessibility/safety point of view (sensory load, disturbing content protection...) it could be viewed as logical to offer an option to hide/deactivate the link previews cards as well, which often contain media out of the reach of moderation tools.

**Comment by quicoto** [comment]
Any thoughts on creating an instance-wide setting for this?

It would also help shave off disk and database space.
